core/templates/dev/head/domain/topic/TopicUpdateService.js

```diff
@@ -210,13 +218,21 @@ oppia.factory('TopicUpdateService', [
           topic.deleteSubtopic(subtopicId, isNewlyCreated);
         }, function(changeDict, topic) {
           // Undo.
-          var subtopicId = _getSubtopicIdFromChangeDict(changeDict);
-          topic.undoDeleteSubtopic(
-            subtopicId, title, skillIds, isNewlyCreated);
-          for (var i = 0; i < skillIds.length; i++) {
-            topic.removeUncategorizedSkillId(skillIds[i]);
-          }
+          throw Error('A deleted subtopic cannot be restored');
         });
+        if (isNewlyCreated) {
```

Comment:
Why is this here?

Do not modify the changeList object on the fly. This should be done during the applyChange() in a way that does not add new fields to the dict.

Also, don't have an "is_deleted" field. Just delete the relevant stuff immediately. Then you don't have to keep track of status of a change cmd over the longer term, that just makes things complicated.

Reply:
So, shall I just delete the change objects here itself without adding the is_deleted field? Since, _applyChange() is common to all functions.

Reply:
I think that deleting the change objects here is the correct thing to do, yes. You might need to bypass the usual "_applyChange" method altogether since you're irreversibly mutating the changelist anyway.

Basically, do whatever modifications you need to do to the changelist here so that applying "undo()" repeatedly all the way to the start of the changelist would still work correctly.

Reply:
Done, removed the 'is_deleted' field and moved the actual deletion of change objects to TopicUpdateService, so everything is done at one place. After which, a setChangeList() function is called in UndoRedoService to just set this modified change list.